File: Self-Organizing Systems/wood-chip-clustering/simulation/clustering_sim.py
import subprocess
import pygame
import numpy as np
import random
import matplotlib.pyplot as plt
import sys

#for capture.
RECORDING = False

#Adjust CELL_SIZE to scale PyGame-Window
CELL_SIZE = 15

#dont modify 
COLOR_BG = (80, 60, 40)
COLOR_OBJECT = (0, 128, 0)
COLOR_AGENT = (0, 255, 0)
COLOR_ANTI_AGENT = (255, 0, 0)
#60Hz max. Probably won't go a lot faster anyways.
SIM_STEPS_PER_SECOND = 1000

#begin modify
PERCENT_ANTI_AGENTS = 0.1
GRID_SIZE = 50
NUM_OBJECTS = 500
NUM_AGENTS = 50
SIM_STEPS = 100000 #10k and 100k respectively
P_PICK = 0.05
P_DROP = 0.95
VISION_RADIUS = 2
ITERRATION = 100
#end modify

#dont change this
NUM_ANTI_AGENTS = int(NUM_AGENTS * PERCENT_ANTI_AGENTS)
WINDOW_SIZE = GRID_SIZE * CELL_SIZE

grid = np.zeros((GRID_SIZE, GRID_SIZE), dtype=int)

object_positions = random.sample([(i, j) for i in range(GRID_SIZE) for j in range(GRID_SIZE)], NUM_OBJECTS)
for x, y in object_positions:
    grid[x, y] = 1

class Agent:
    p_density = 0.15
    p_pick = P_PICK
    p_drop = P_DROP
    vision_r = VISION_RADIUS
    carrying = False
    
    def __init__(self, is_anti=False):
        self.x = random.randint(0, GRID_SIZE - 1)
        self.y = random.randint(0, GRID_SIZE - 1)
        self.is_anti = is_anti

    #move robot. we dont care about collision. Ants also crawl over each other :D and im lazy.
    #only use random movement.
    def move(self):
        #only move left/right
        dx, dy = random.choice([(0,1), (0,-1), (1,0), (-1,0)])

        #python % operator actually makes sense so we can just use it. -1 % 3 = 2. Easy wrap around.
        self.x = (self.x + dx) % GRID_SIZE
        self.y = (self.y + dy) % GRID_SIZE
        return
        #non random movement makes things worse actually..... maybe because anti-agent have a different logic here...
        #we dont want seperate logics, just seperate probabilities.
        '''
        best_dx, best_dy = 0, 0
        max_count = -1

        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                if dx == 0 and dy == 0:
                    continue

                nx = (self.x + dx) % GRID_SIZE
                ny = (self.y + dy) % GRID_SIZE

                count = 0
                for i in [-1, 0, 1]:
                    for j in [-1, 0, 1]:
                        lx = (nx + i) % GRID_SIZE
                        ly = (ny + j) % GRID_SIZE
                        if grid[lx][ly] == 1:
                            count += 1

                if count > max_count:
                    max_count = count
                    best_dx = dx
                    best_dy = dy

        if max_count > 0 and random.randint(0, 99) < 70:
            self.x = (self.x + best_dx) % GRID_SIZE
            self.y = (self.y + best_dy) % GRID_SIZE
        else:
            moves = [(0,1), (0,-1), (1,0), (-1,0)]
            dx, dy = random.choice(moves)
            self.x = (self.x + dx) % GRID_SIZE
            self.y = (self.y + dy) % GRID_SIZE
        '''

    #look at all cells in vision range and check if they are populated.
    def local_density(self):
        grid_size = grid.shape[0]
        count = 0
        total_cells = (2 * self.vision_r + 1)**2 - 1

        for dx in range(-self.vision_r, self.vision_r + 1):
            for dy in range(-self.vision_r, self.vision_r + 1):
                nx = (self.x + dx) % grid_size
                ny = (self.y + dy) % grid_size
                if grid[nx, ny] == 1:
                    count += 1

        return count / total_cells
    
    #just some sugar
    def try_pick_up(self):
        if (grid[self.x, self.y] == 1) and (not self.carrying):
            grid[self.x, self.y] = 0
            self.carrying = True

    #just some sugar
    def try_drop(self):
        if (grid[self.x, self.y] == 0) and (self.carrying):
            grid[self.x, self.y] = 1
            self.carrying = False

    #just some sugar
    def pick_probability(self,density):
        return (self.p_pick / (self.p_pick + density)) ** 2

    #just some sugar
    def drop_probability(self,density):
        return (density / (self.p_drop + density)) ** 2

    #should i pick up or drop asked the ant to itself.
    def act(self):
        density = self.local_density()
            
        if not self.carrying:
            prob = self.pick_probability(density)
            if self.is_anti:
                prob = 1 - prob
            if random.random() < prob:
                self.try_pick_up()
        else:
            prob = self.drop_probability(density)
            if self.is_anti:
                prob = 1 - prob
            if random.random() < prob:
                self.try_drop()

#populate agents array. Total agents = ANTI+NORMAL
agents = [Agent(False) for _ in range(NUM_AGENTS)] + [Agent(True) for _ in range(NUM_ANTI_AGENTS)]

def draw_grid(screen, grid, agents, draw_vision):
    screen.fill(COLOR_BG)
    for x in range(GRID_SIZE):
        for y in range(GRID_SIZE):
            if grid[x, y] == 1:
                pygame.draw.rect(screen, COLOR_OBJECT,
                                 (y * CELL_SIZE, x * CELL_SIZE, CELL_SIZE, CELL_SIZE))

    for agent in agents:
        color = COLOR_ANTI_AGENT if agent.is_anti else COLOR_AGENT
        pygame.draw.circle(screen, color,
                           (agent.y * CELL_SIZE + CELL_SIZE // 2, agent.x * CELL_SIZE + CELL_SIZE // 2),
                           CELL_SIZE // 2)
        if draw_vision:
            pygame.draw.circle(screen, (0, 0, 0), (agent.y * CELL_SIZE + CELL_SIZE // 2, agent.x * CELL_SIZE + CELL_SIZE // 2), agent.vision_r * CELL_SIZE,
                           2)

    #only update the portion needed to update.
    #might result in weird artefacts but its faster.
    pygame.display.flip()

def simulate(run_visualization=True, draw_vision=True):
    if run_visualization:
        pygame.init()
        screen = pygame.display.set_mode((WINDOW_SIZE, WINDOW_SIZE))
        pygame.display.set_caption("Wood Chip Clustering Simulation")
        clock = pygame.time.Clock()

    for step in range(SIM_STEPS):
        print(step)

        #pygame exit handler.
        for event in pygame.event.get() if run_visualization else []:
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

        for agent in agents:
            agent.move()
            agent.act()

        if run_visualization:
            draw_grid(screen, grid, agents, draw_vision)
            #limit frame rate. like a wait but from pygames.
            clock.tick(SIM_STEPS_PER_SECOND)

    if run_visualization:
        while(RECORDING):
            pass
        pygame.quit()

    return np.argwhere(grid == 1)

def plot_final_objects(initial_positions, final_positions):
    #x_i, y_i = zip(*initial_positions)
    x, y = zip(*final_positions)
    plt.figure(figsize=(6,6))
    #plt.scatter(y_i, x_i, s=10, c='blue')
    plt.scatter(y, x, s=10, c='red')
    plt.title("Final Object Clustering")
    #since we use PyGame data. Flip it all (0,0) is top left corner.
    plt.gca().invert_yaxis()
    plt.savefig("final_clustering.png")

#call c-script from python
def run_c_sim(p_pick, p_drop, num_agents, percent_anti_agents, sim_steps, density_radius, grid_size, num_obj, iter):
    result = subprocess.run(
            ["./clustering_sim",str(p_pick), str(p_drop), str(num_agents), str(percent_anti_agents), str(sim_steps), str(density_radius), str(grid_size), str(num_obj), str(iter)],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            check=True
        )
    return float(result.stdout.strip().split('\n')[0])

def run_animated_simulation(run_visualization, draw_vision):
    #modify at top of file.
    initial_positions = np.argwhere(grid == 1)
    final_positions = simulate(run_visualization=run_visualization, draw_vision=draw_vision)
    #initial position is uncommented in the script, add back in to see starting conditions.
    plot_final_objects(initial_positions, final_positions)

def sweep_anti_agent_percentage():
    cluster_sizes = []
    percent_anti_agents_values = [0, 0.05, 0.1, 0.15]

    for p in percent_anti_agents_values:
        print(f"percent_anti_agents = {p:.3f}")
        c_size = run_c_sim(P_PICK, P_DROP, NUM_AGENTS, p, SIM_STEPS, VISION_RADIUS, GRID_SIZE, NUM_OBJECTS, ITERRATION)
        if c_size is None:
            c_size = 0
        cluster_sizes.append(c_size)

    plt.figure(figsize=(8,5))
    plt.plot(percent_anti_agents_values, cluster_sizes, marker='o')
    plt.xlabel('Percent Anti Agents')
    plt.ylabel('Cluster Size')
    plt.title('Cluster Size vs Percent Anti Agents')
    plt.grid(True)
    plt.savefig("cluster_size_vs_anti_agent_percent.png")
    
run_animated_simulation(run_visualization = True, draw_vision = True)
sweep_anti_agent_percentage()

File: Self-Organizing Systems/wood-chip-clustering/simulation/clustering_sim.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>
#include <math.h>

#define PYTHON_EQUIVALENT
#define CLUSTER_DIST (2)

/**
 * 
 * No need to look at this.
 * 1:1 Translation of the Python-Code.
 */

//2D Array [x,y]
static int **grid;
static int GRID_SIZE = 100;
static int NUM_OBJECTS = 100;
static int VISION_R = 5;
static double P_PICK = 0;
static double P_DROP = 0;

//Base Probability is not reversed. Reversed after calculation and before logic! (1- density_prob)
typedef struct {
    int x, y;
    int is_anti;
    double p_pick;
    double p_drop;
    int carrying;
} Agent;

//just some sugar
double pick_probability(double p_pick, double density) {
    return pow(p_pick / (p_pick + density), 2);
}

double drop_probability(double p_drop, double density) {
    return pow(density / (p_drop + density), 2);
}

void try_pick_up(Agent* agent) {
    if (grid[agent->x][agent->y] == 1 && !agent->carrying) {
        grid[agent->x][agent->y] = 0;
        agent->carrying = 1;
    }
}

void try_drop(Agent* agent) {
    if (grid[agent->x][agent->y] == 0 && agent->carrying) {
        grid[agent->x][agent->y] = 1;
        agent->carrying = 0;
    }
}

//python modulo
int mod_wrap_remaineder(int a, int b) {
    int r = a % b;
    // -3 % 2 = -1 -> -1 + 2 = 1
    return r < 0 ? r + b : r;
}

double local_density(int x, int y, int radius) {
    int count = 0;
    int total_cells = (2 * radius + 1) * (2 * radius + 1) - 1;

    for (int dx = -radius; dx <= radius; dx++) {
        for (int dy = -radius; dy <= radius; dy++) {
            int nx = mod_wrap_remaineder(x + dx, GRID_SIZE);
            int ny = mod_wrap_remaineder(y + dy, GRID_SIZE);
            if (grid[nx][ny] == 1) {
                count++;
            }
        }
    }

    return (double)count / total_cells;
}

void move_agent(Agent* agent) {
#ifdef PYTHON_EQUIVALENT
    int moves[4][2] = {{0,1},{0,-1},{1,0},{-1,0}};
    int r = rand() % 4;
    agent->x = mod_wrap_remaineder(agent->x + moves[r][0], GRID_SIZE);
    agent->y = mod_wrap_remaineder(agent->y + moves[r][1], GRID_SIZE);
#else
    int best_dx = 0, best_dy = 0;
    int max_count = -1;

    for (int dx = -1; dx <= 1; dx++) {
        for (int dy = -1; dy <= 1; dy++) {
            if (dx == 0 && dy == 0) continue;

            int nx = mod_wrap_remaineder(agent->x + dx, GRID_SIZE);
            int ny = mod_wrap_remaineder(agent->y + dy, GRID_SIZE);

            int count = 0;
            for (int i = -1; i <= 1; i++) {
                for (int j = -1; j <= 1; j++) {
                    int lx = mod_wrap_remaineder(nx + i, GRID_SIZE);
                    int ly = mod_wrap_remaineder(ny + j, GRID_SIZE);
                    if (grid[lx][ly] == 1) count++;
                }
            }

            if (count > max_count) {
                max_count = count;
                best_dx = dx;
                best_dy = dy;
            }
        }
    }

    if (max_count > 0 && (rand() % 100 < 70)) {
        agent->x = mod_wrap_remaineder(agent->x + best_dx, GRID_SIZE);
        agent->y = mod_wrap_remaineder(agent->y + best_dy, GRID_SIZE);
    } else {
        int moves[4][2] = {{0,1},{0,-1},{1,0},{-1,0}};
        int r = rand() % 4;
        agent->x = mod_wrap_remaineder(agent->x + moves[r][0], GRID_SIZE);
        agent->y = mod_wrap_remaineder(agent->y + moves[r][1], GRID_SIZE);
    }
#endif
}

//bfs joinked from stack overflow
int bfs_cluster(int start_x, int start_y, bool** visited, int max_empty) {
    int dx[] = {0, 0, -1, 1};
    int dy[] = {-1, 1, 0, 0};
    int queue_size = GRID_SIZE * GRID_SIZE;
    int queue_head = 0, queue_tail = 0;
    int cluster_size = 0;

    // (x, y, empty_used)
    int (*queue)[3] = malloc(queue_size * sizeof(int[3])); 

    queue[queue_tail][0] = start_x;
    queue[queue_tail][1] = start_y;
    queue[queue_tail][2] = 0;
    queue_tail++;

    int** local_visited = malloc(GRID_SIZE * sizeof(int*));
    for (int i = 0; i < GRID_SIZE; i++) {
        local_visited[i] = malloc(GRID_SIZE * sizeof(int));
        for (int j = 0; j < GRID_SIZE; j++)
            local_visited[i][j] = -1;
    }
    local_visited[start_x][start_y] = 0;

    while (queue_head < queue_tail) {
        int x = queue[queue_head][0];
        int y = queue[queue_head][1];
        int empty_used = queue[queue_head][2];
        queue_head++;

        if (visited[x][y]) continue;
        visited[x][y] = true;
        if (grid[x][y] == 1) cluster_size++;

        for (int i = 0; i < 4; i++) {
            int nx = mod_wrap_remaineder(x + dx[i], GRID_SIZE);
            int ny = mod_wrap_remaineder(y + dy[i], GRID_SIZE);
            int new_empty = empty_used + (grid[nx][ny] == 0 ? 1 : 0);

            if (new_empty > max_empty) continue;
            if (local_visited[nx][ny] != -1 && local_visited[nx][ny] <= new_empty) continue;

            local_visited[nx][ny] = new_empty;
            queue[queue_tail][0] = nx;
            queue[queue_tail][1] = ny;
            queue[queue_tail][2] = new_empty;
            queue_tail++;
        }
    }

    for (int i = 0; i < GRID_SIZE; i++)
        free(local_visited[i]);
    free(local_visited);
    free(queue);

    return cluster_size;
}

int find_largest_cluster_size(int empty_tiles) {
    bool** visited = malloc(GRID_SIZE * sizeof(bool*));
    for (int i = 0; i < GRID_SIZE; i++) {
        visited[i] = malloc(GRID_SIZE * sizeof(bool));
        for (int j = 0; j < GRID_SIZE; j++)
            visited[i][j] = false;
    }

    int max_cluster = 0;
    for (int x = 0; x < GRID_SIZE; x++) {
        for (int y = 0; y < GRID_SIZE; y++) {
            if (!visited[x][y] && grid[x][y] == 1) {
                int size = bfs_cluster(x, y, visited, empty_tiles);
                if (size > max_cluster)
                    max_cluster = size;
            }
        }
    }
    for (int i = 0; i < GRID_SIZE; i++)
        free(visited[i]);
    free(visited);

    return max_cluster;
}

double run_simulation(int num_agents, double percent_anti_agents, int sim_steps) {
    for (int i=0; i<GRID_SIZE; i++)
        for (int j=0; j<GRID_SIZE; j++)
            grid[i][j] = 0;

    srand((unsigned int)time(NULL));

    int total_cells = GRID_SIZE * GRID_SIZE;
    int* indices = (int*)malloc(total_cells * sizeof(int));
    for (int i=0; i<total_cells; i++) indices[i] = i;

    for (int i=0; i<NUM_OBJECTS; i++) {
        int r = i + rand() % (total_cells - i);
        int tmp = indices[i]; indices[i] = indices[r]; indices[r] = tmp;

        int pos = indices[i];
        int x = pos / GRID_SIZE;
        int y = pos % GRID_SIZE;
        grid[x][y] = 1;
    }
    free(indices);

    int num_anti_agents = (int)(num_agents * percent_anti_agents);
    int normal_agents = num_agents - num_anti_agents;
    Agent* agents = (Agent*)malloc(num_agents * sizeof(Agent));

    for (int i=0; i<num_agents; i++) {
        agents[i].x = rand() % GRID_SIZE;
        agents[i].y = rand() % GRID_SIZE;
        agents[i].carrying = 0;
        agents[i].is_anti = (i >= normal_agents) ? 1 : 0;
        agents[i].p_pick = P_PICK;
        agents[i].p_drop = P_DROP;
    }

    for (int step=0; step<sim_steps; step++) {
        for (int i=0; i<num_agents; i++) {
            Agent* a = &agents[i];
            move_agent(a);

            double density = local_density(a->x, a->y, VISION_R);
            if (!a->carrying) {
                double prob = pick_probability(a->p_pick, density);
                if(a->is_anti){
                    prob = 1 - prob;
                }
                if (((double)rand() / RAND_MAX) < prob) {
                    try_pick_up(a);
                }
            } else {
                double prob = drop_probability(a->p_drop, density);
                if(a->is_anti){
                    prob = 1 - prob;
                }
                if (((double)rand() / RAND_MAX) < prob) {
                    try_drop(a);
                }
            }
        }
    }

    free(agents);

    //replace this with loop using local_density for alternative to cluster size.
    return find_largest_cluster_size(CLUSTER_DIST);
}

int main(int argc, char** argv) {
    if (argc < 10) {
        fprintf(stderr, "Usage: p_pick p_drop num_agents percent_anti_agents sim_steps density_radius grid_size num_objects iterations\n");
        return 1;
    }

    P_PICK = atof(argv[1]);
    P_DROP = atof(argv[2]);
    int num_agents = atoi(argv[3]);
    double percent_anti_agents = atof(argv[4]);
    int sim_steps = atoi(argv[5]);
    VISION_R = atoi(argv[6]);
    GRID_SIZE = atoi(argv[7]);
    NUM_OBJECTS = atoi(argv[8]);
    int iter      = atoi(argv[9]);
    
    grid = malloc(GRID_SIZE * sizeof(int *));
    for (int i = 0; i < GRID_SIZE; i++) {
        grid[i] = malloc(GRID_SIZE * sizeof(int));
    }

    double avg = 0;
    for(int i=0; i <iter; i++){
        avg +=run_simulation(num_agents, percent_anti_agents, sim_steps);
    }
    printf("%f", avg/iter);
    return 0;
}
